### ISSUE TITLE
CONNECT to https doesn't sniff Host in TLS/SSL negotiation

### ISSUE THREAD

**Issue by abourget** [elazarl/goproxy#86]
I see `CONNECT ip:443` going through, and the name of the host the request is destined to in the next few bytes.. however, the current implementation signs the mitm cert with the IP in the `CONNECT` line, and doesn't allow us to direct the traffic to the right place, according to host-based rules.

Would it be possible to sniff the first packet after the connect, and use that info to determine the *real* host we're talking about.. and put back the bytes in a buffer so that further processing can be done by the right libs (like TLS taking the lead from before our read).

Not sure it's really clear.. but does no one find awkward we only get IPs to filter on HTTPS connect ?

**Comment by abourget** [comment]
I think I see.. I'm using Android's `adb` parameter `--http-proxy`, and it would seem that it is done at the emulator level, so it can't know the host the application asked.. only the IP that resulted. Since it's not an application running on Android, and asking for a domain, I only get an already resolved domain name.

Firefox on my desktop, issues a CONNECT with a domain name.. giving me the freedom to do the DNS resolution, and to discriminate requests based on the domain.

It would be *realllly* cool to sniff the SNI field in the incoming TLS negotiation, and provide that as a separate condition perhaps.. and loop it back through the OnRequest() handlers.

**Comment by elazarl** [comment]
I wanted to do that, but I saw no way of getting SNI info from the standard
library.

If you have an elegant way, I'll be happy to hear about it.

**Comment by mzimmerman** [comment]
I found an [excellent SNI sniffing library] (github.com/inconshreveable/go-vhost) that I've used in [github.com/mzimmerman/whitelistproxy](https://github.com/mzimmerman/whitelistproxy/blob/master/transparent.go#L441) -- I'm not sure how well it would fit into a generic goproxy implementation.  I even hacked support for non-SNI enabled clients by guessing they were visiting the site of the most recent DNS lookup (this assumes the goproxy server is also the DNS server)

**Comment by elazarl** [comment]
It's probably a good idea to use it, although I really liked the idea
goproxy have no external deps.

I'll try to look into it in a few weeks.

**Comment by abourget** [comment]
yeah that whitelistproxy is exactly what I had in mind.  Where would we put that ?  Would we make a `ReqCond` to enable SNI only on IP-based CONNECTs, and otherwise go straight through with the provided host.. if we get an IP on CONNECT, then go through the whitelistproxy example, and re-feed the socket back to our https handlers ?

Maybe we could add a flag to `ProxyHttpServer` to allow disabling/enabling of such a feature?

For dependencies.. I don,t see a huge problem, it's always hard to break the mold.. but I really feel goproxy could replace many uses of Charles Proxy, mitmproxy and Fiddler .. in a more coder-friendly and performant way.

What do you think ?

**Comment by johnweldon** [comment]
@elazarl why no external deps?  1) possibility of api changes or breakage or disappearing? or 2) you want all the code to be directly under the control of this repo?

To fix 1 you can either use gopkg.in or fork the dependency or both, or pull your dependencies into a local sub folder.

To fix 2 you'd just have to re-implement SNI in goproxy :) (or some variation on the solution to 1 might work too)

**Comment by elazarl** [comment]
Having zero external dependencies makes development and usage easier. For
example if someone wants to embed goproxy in his code - he needs no other
distribution.

But I already agreed SNI is a good idea, and it doesn't look like the
standard library would support it any time soon. I'll indeed add support
for it when I'll have time. Hopefully sooner than later.

**Comment by johnweldon** [comment]
Thanks @elazarl

**Comment by abourget** [comment]
@elazarl if you can think of a neat way to wire it in, I'm willing to do the work.. do you have anything in mind ?

**Comment by abourget** [comment]
https://github.com/elazarl/goproxy/pull/97
